### PR TITLE
logic that does pytest discovery

### PR DESCRIPTION
**Pull request by eleanorjboyd**

### CODE REVIEW

File: pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        if type(test_case.parent) == pytest.Module:
+            file_nodes_dict.setdefault(
+                test_case.parent, create_file_node(test_case.parent)
+            )["children"].append(test_node)
+        else:
+            test_class_node = class_nodes_dict.setdefault(
+                test_case.parent.name,
+                create_class_node(test_case.parent),
+            )
+            test_class_node["children"].append(test_node)
```
Comment (brettcannon) [inline pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py:76]: How is this code different than the lines above? Is it just the dict you are working with? If so then I would assign to a new variable that is assigned based on the conditional guard and then uses the same code.

File: pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                test_file_node["children"].append(test_class_node)
+
+    created_files_folders_dict: dict[str, TestNode] = {}
+    for file_module, file_node in file_nodes_dict.items():
+        root_folder_node = build_nested_folders(
+            file_module, file_node, created_files_folders_dict, session
+        )
+        # the final folder we get to is the highest folder in the path and therefore we add this as a child to the session.
+        if root_folder_node.get("id_") not in session_children_dict:
+            session_children_dict[root_folder_node.get("id_")] = root_folder_node
```
Comment (brettcannon) [inline pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py:93]: Should `None` be used as a key in `session_children_dict`, as `dict.get()` will return that if it isn't there? And what if "id_" isn't there? Is `None` a legitimate key?

And assuming `None` is what you want if `"id_"` doesn't exist, then I would not get it twice:

```suggestion
        root_folder_id = root_folder_node.get("id_")
        if root_folder_id not in session_children_dict:
            session_children_dict[root_folder_id] = root_folder_node
```

File: pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    session_node["children"] = list(session_children_dict.values())
+    return session_node, errors
+
+
+def build_nested_folders(
+    file_module, file_node, created_files_folders_dict, session
```
Comment (brettcannon) [inline pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py:99]: Type the parameters (and everywhere else).

File: pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    status: Literal["success", "error"]
+    tests: NotRequired[TestNode]
+    errors: NotRequired[List[str]]
+
+
+def sendPost(cwd, tests):
+    payload: PayloadDict = {"cwd": cwd, "status": "success", "tests": tests}
+    testPort = os.getenv("TEST_PORT", 45454)
+    testuuid = os.getenv("TEST_UUID")
+    addr = ("localhost", int(testPort))
```
Comment (brettcannon) [inline pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py:178]:
```suggestion
    addr = "localhost", int(testPort)
```

File: pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+    print("sending post", addr, cwd)
```
Comment (brettcannon) [inline pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py:179]: Is this purposefully printed out all the time? Or was this debugging code?

File: pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    with socket_manager.SocketManager(addr) as s:
+        data = json.dumps(payload)
+        request = f"""POST / HTTP/1.1
+Host: localhost:{testPort}
+Content-Length: {len(data)}
+Content-Type: application/json
+Request-uuid: {testuuid}
+
+{data}"""
+        result = s.socket.sendall(request.encode("utf-8"))  # type: ignore
```
Comment (brettcannon) [inline pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py:189]: That socket only needs to be open for the send, so you should keep it open for a shorter amount of time. You are also assigning to a variable and then never using it.

```suggestion
    
    data = json.dumps(payload)
    request = f"""POST / HTTP/1.1
Host: localhost:{testPort}
Content-Length: {len(data)}
Content-Type: application/json
Request-uuid: {testuuid}

{data}"""
     with socket_manager.SocketManager(addr) as s:
        s.socket.sendall(request.encode("utf-8"))  # type: ignore
```

Comment (brettcannon) [reply pythonFiles/pytest-vscode-integration/pytest_vscode_integration.py:189]: And why the `type: ignore` opt-out?

**Review by brettcannon — CHANGES_REQUESTED** [review]

**Review by sophacola — APPROVED** [review]

**Pull request closed without merging**